Steve/rapidprototype

stevenkboyd opened the pull request:
I've done an extensive overhaul of blRapidPrototype to provide the ability to manipulate STL files. This was necessary for our TBone project that combined bone models with other geometric features. It is now possible to create some basic objects (cylinder, cube, etc), generate an STL file from AIM (as usual), generate an AIM from STL (new), and boolean operations of union, difference and intersection of STL files. This will be very useful for creating custom models in the future.

Besler commented on bonelab/cli/RapidPrototypeLetters.py:
```diff
@@ -0,0 +1,487 @@
+  mat4x4 = visualize_actors( cube_and_posts.GetOutputPort(), bone1234.GetOutputPort()  )
+  #mat4x4 = visualize_actors( cube_and_posts.GetOutputPort(), None  )
+  
+  #ofile = "/Users/skboyd/Documents/projects/FormLabs/radius_time_series/bone_mask_post_and_table.stl" 
```
You could remove the path or remove all the comments.
```suggestion
  #ofile = "bone_mask_post_and_table.stl" 
```

Besler commented on bonelab/cli/RapidPrototypeLetters.py:
```diff
@@ -0,0 +1,487 @@
+  writer.SetFileName(ofile)
+  writer.Update()
+  
+  exit()
```
Best to use `return` instead of `exit` in a function
```suggestion
  return
```

Besler commented on bonelab/cli/RapidPrototypeLetters.py:
```diff
@@ -0,0 +1,487 @@
+    message('ERROR: User can scale width or height, but not both.',
+            '{:8s} = {:8.2f} mm'.format('width',width),
+            '{:8s} = {:8.2f} mm'.format('height',height))
+    exit()
```
Probably `return` here?
```suggestion
    return
```

Besler commented on bonelab/cli/RapidPrototypeLetters.py:
```diff
@@ -0,0 +1,487 @@
+  else:
+    message('No output file designated. No file written.')
+
+  exit()
```
Function call again
```suggestion
  return
```

Besler commented on bonelab/cli/RapidPrototypeLetters.py:
```diff
@@ -0,0 +1,487 @@
+        prog="blRapidPrototypeLetters",
+        description=description
+    )
+    #parser.add_argument('input_filename', help='Input image file')
```
Should this line be commented out?

Besler commented on bonelab/cli/RapidPrototypeLetters.py:
```diff
@@ -0,0 +1,487 @@
+
+    # Run program
+    CreateShapes(**vars(args))
+    #CreateTimeSeries()
```
Remove if not used

Besler commented on bonelab/cli/RapidPrototype.py:
```diff
@@ -3,158 +3,978 @@
+    try:    
+       mat4x4 = readTransform(transform)
+    except OSError:
+       print("ERROR: File could not be read: " + transform)
+       exit(0)
```
Do you actually want to exit here? A more typical solution would be to throw an exception.
```suggestion
       raise SystemError('File could not be read: ' + transform)
```

Besler commented on bonelab/cli/RapidPrototype.py:
```diff
@@ -3,158 +3,978 @@
+  elif "difference" in operation:
+    booleanOperation.SetOperationToDifference()
+  else:
+    message("ERROR: Invalid boolean operation.")
+    exit()
```
Again, probably don't want to exit here. Instead, throw error or return.

```suggestion
    raise ValueError('Invalid boolean operation: ' + operation)
```

Besler commented:
See the minimal testing at [test_cli_setup.py](https://github.com/Bonelab/Bonelab/blob/master/tests/cli/test_cli_setup.py). You'd be writing a function like:

```python
def test_ blRapidPrototypeLetters(self):
        '''Can run `blRapidPrototypeLetters `'''
        self.runner('blRapidPrototypeLetters')
```

stevenkboyd commented:
Hi Bryce,

The program “RapidPrototypeLetters.py” should actually have been deleted as it’s no longer required. All the functionality is built into “RapidPrototype.py”. I’ve updated the branch now and fixed some of the issues you raised.

I’ve also added a test in test_cli_setup.py as well as a set of tests defined in “test_blRapidPrototype.py”. I had to add some files into /testing/data for those tests, and I’m a little unclear what repository those get stored in. Is it the BonelabData repository? The files I created are:

cube_expected.stl
cylinder_expected.stl
sign_expected.stl
sphere_expected.stl
test25a_expected.stl

If you could check that the tests are working, then I’m hoping to merge the branch into the master.

Incidentally, I tried to make a test in “test_blRapidPrototype.py” that would test ‘stl2aim’, but I couldn’t make it work. It seemed to fail upon trying to read the AIM file I had created, which is strange. Not a big deal, and I guess we can live without perfect testing.

Cheers,

Steve

--
Steven Boyd
PhD, Professor, University of Calgary
Director, McCaig Institute for Bone and Joint Health
Full Signature<https://wpsites.ucalgary.ca/steven-boyd/>

On Jun 5, 2020, at 10:02 AM, Bryce Besler <notifications@github.com<mailto:notifications@github.com>> wrote:

[△EXTERNAL]




See the minimal testing at test_cli_setup.py<https://github.com/Bonelab/Bonelab/blob/master/tests/cli/test_cli_setup.py>. You'd be writing a function like:

def test_ blRapidPrototypeLetters(self):
        '''Can run `blRapidPrototypeLetters `'''
        self.runner('blRapidPrototypeLetters')

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/Bonelab/Bonelab/pull/13#issuecomment-639596198>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFAQITQTUY2H3PFMC7A4UZ3RVEJHVANCNFSM4NS3JYCQ>.

Besler commented:
All looks good! In the future we should create a directory for testing data outside the repository. The STL files are small, but a commit creates a save of all files each time, which can blow up for large data types. Looks good!
